File: week1/credit.c
#include <cs50.h>
#include <stdio.h>

#define AMEX_PREFIX1 34
#define AMEX_PREFIX2 37
#define AMEX_LENGTH 15
#define MASTERCARD_PREFIX_START 51
#define MASTERCARD_PREFIX_END 55
#define MASTERCARD_LENGTH 16
#define VISA_PREFIX 4
#define VISA_LENGTH1 13
#define VISA_LENGTH2 16

string card_type(long n);
bool valid_card(long n);

int main(void)
{
    long card_number = get_long("Enter card number: ");

    if (valid_card(card_number))
    {
        printf("%s\n", card_type(card_number));
    }
    else
    {
        printf("INVALID\n");
    }
}

string card_type(long n)
{
    long temp = n;
    int card_length = 2;

    while (temp >= 100)
    {
        temp /= 10;
        card_length++;
    }

    if ((temp == AMEX_PREFIX1 || temp == AMEX_PREFIX2) && card_length == AMEX_LENGTH)
    {
        return "AMEX";
    }
    else if ((temp >= MASTERCARD_PREFIX_START && temp <= MASTERCARD_PREFIX_END) &&
             card_length == MASTERCARD_LENGTH)
    {
        return "MASTERCARD";
    }
    // one less for VISA
    temp /= 10;
    if (temp == VISA_PREFIX && (card_length == VISA_LENGTH1 || card_length == VISA_LENGTH2))
    {
        return "VISA";
    }

    return "INVALID";
}

// Use Luhn's algorithm to check card checksum
bool valid_card(long n)
{
    int sum = 0;
    long temp = n;
    bool alternate = false;

    while (temp > 0)
    {
        // gets last digit
        int digit = temp % 10;
        if (alternate)
        {
            digit *= 2;
            // when over 10 need to 'add' digits together e.g. 15 -> 6, 18 -> 9
            sum += (digit > 9) ? (digit - 9) : digit;
        }
        else
        {
            sum += digit;
        }

        alternate = !alternate;
        // removes last digit
        temp /= 10;
    }

    // Checks if ends in 0
    return (sum % 10 == 0);
}

File: week6/sentimental-credit/credit.py
import re

AMEX_PREFIX1 = 34
AMEX_PREFIX2 = 37
AMEX_LENGTH = 15
MASTERCARD_PREFIX_START = 51
MASTERCARD_PREFIX_END = 55
MASTERCARD_LENGTH = 16
VISA_PREFIX = 4
VISA_LENGTH1 = 13
VISA_LENGTH2 = 16

# try:
#     card_no = int(input("Number: "))
# except ValueError:
#     print("Only enter numbers between 13 and 16 digits long")
#     exit()

def get_input():
    card_input = input("Number: ").strip()
    # would check if 51-55, 34, 37 or 4 start then 11-14 extra digits so 12-16 long
    # if not re.fullmatch(r"(5[1-5]|34|37|4)\d{11,14}", card_input):
    # regexp to check 13-16 long
    if not re.fullmatch(r"\d{13,16}", card_input):
        print("Invalid input: must start with 51–55, 34, 37, or 4 and be 13–16 digits")
        exit()
    return card_input

def card_type(n):
    prefix = int(n[:2])
    v_prefix = int(n[0])
    card_length = len(n)

    if ((prefix == AMEX_PREFIX1 or prefix == AMEX_PREFIX2) and card_length == AMEX_LENGTH):
        return "AMEX"
    elif ((prefix >= MASTERCARD_PREFIX_START and prefix <= MASTERCARD_PREFIX_END) and
          card_length == MASTERCARD_LENGTH):
        return "MASTERCARD"
    elif (v_prefix == VISA_PREFIX and (card_length == VISA_LENGTH1 or card_length == VISA_LENGTH2)):
        return "VISA"

    return "INVALID"

# Luhn's algorithm to check card checksum
def valid_card(n):
    sum = 0
    alternate = False

    while (n > 0):
        # get last digit
        digit = n % 10
        if (alternate):
            digit *= 2
            # add together when over 10, 15 -> 6
            sum += digit - 9 if digit > 9 else digit
        else:
            sum += digit

        alternate = not alternate
        n = n // 10

    return sum % 10 == 0

card_no = get_input()
# turn string into int for valid_card
if (valid_card(int(card_no))):
    print(f"{card_type((card_no))}")
else:
    print("INVALID")
